In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt

import numpy as np
import pandas as pd
import scipy.stats

import seaborn as sns

# Reading and examining the datasets

Checking the hibernation data and loading it into a dataframe.

In [2]:
!head -n 5 'data/HibernationSurveyObservations.csv'

observationID,recordKey,organisationName,datasetKey,surveyKey,sampleKey,gridReference,precision,siteKey,siteName,featureKey,startDate,endDate,dateType,recorder,determiner,pTaxonVersionKey,taxonName,authority,commonName,taxonGroup,sensitive,zeroAbundance,fullVersion
463166539,1767,The Bat Conservation Trust,GA000612,1,1,SM90,10km,,,150161,12/02/2006,12/02/2006,Day,,,NBNSYS0100004720,Pipistrellus pygmaeus,"(Leach, 1825)",Soprano Pipistrelle,terrestrial mammal,false,false,false
463166623,4325,The Bat Conservation Trust,GA000612,1,1,TL65,10km,,,168651,21/12/2008,21/12/2008,Day,,,NHMSYS0020001355,Pipistrellus pipistrellus sensu stricto,"(Schreber, 1774)",Common Pipistrelle,terrestrial mammal,false,false,false
463166624,4326,The Bat Conservation Trust,GA000612,1,1,TL65,10km,,,168651,18/01/2009,18/01/2009,Day,,,NHMSYS0020001355,Pipistrellus pipistrellus sensu stricto,"(Schreber, 1774)",Common Pipistrelle,terrestrial mammal,false,false,false
463166625,4327,The Bat Conservation Trust,GA000612,1

In [3]:
!tail -n 5 'data/HibernationSurveyObservations.csv'

463184995,20315,The Bat Conservation Trust,GA000612,1,1,TQ35,10km,,,160693,20/01/2013,20/01/2013,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false
463185027,16096,The Bat Conservation Trust,GA000612,1,1,TQ21,10km,,,162216,08/01/2005,08/01/2005,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false
463185029,16098,The Bat Conservation Trust,GA000612,1,1,TQ21,10km,,,162216,20/02/2005,20/02/2005,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false
463185031,16103,The Bat Conservation Trust,GA000612,1,1,TQ31,10km,,,165960,20/02/2005,20/02/2005,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false
463185032,17955,The Bat Conservation Trust,GA000612,1,1,TQ31,10km,,,165960,03/02/2008,03/02/2008,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,

In [4]:
hibernation_df = pd.read_csv('data/HibernationSurveyObservations.csv', engine='python')

In [5]:
hibernation_df.head()

observationID  recordKey            organisationName datasetKey  surveyKey  \
0      463166539       1767  The Bat Conservation Trust   GA000612          1   
1      463166623       4325  The Bat Conservation Trust   GA000612          1   
2      463166624       4326  The Bat Conservation Trust   GA000612          1   
3      463166625       4327  The Bat Conservation Trust   GA000612          1   
4      463166626       4328  The Bat Conservation Trust   GA000612          1   

   sampleKey gridReference precision  siteKey  siteName     ...      recorder  \
0          1          SM90      10km      NaN       NaN     ...           NaN   
1          1          TL65      10km      NaN       NaN     ...           NaN   
2          1          TL65      10km      NaN       NaN     ...           NaN   
3          1          TL65      10km      NaN       NaN     ...           NaN   
4          1          TL65      10km      NaN       NaN     ...           NaN   

  determiner  pTaxonVersionKey                                taxonName  \
0        NaN  NBNSYS0100004720                    Pipistrellus pygmaeus   
1        NaN  NHMSYS0020001355  Pipistrellus pipistrellus sensu stricto   
2        NaN  NHMSYS0020001355  Pipistrellus pipistrellus sensu stricto   
3        NaN  NHMSYS0020001355  Pipistrellus pipistrellus sensu stricto   
4        NaN  NHMSYS0020001355  Pipistrellus pipistrellus sensu stricto   

          authority           commonName          taxonGroup sensitive  \
0     (Leach, 1825)  Soprano Pipistrelle  terrestrial mammal     False   
1  (Schreber, 1774)   Common Pipistrelle  terrestrial mammal     False   
2  (Schreber, 1774)   Common Pipistrelle  terrestrial mammal     False   
3  (Schreber, 1774)   Common Pipistrelle  terrestrial mammal     False   
4  (Schreber, 1774)   Common Pipistrelle  terrestrial mammal     False   

  zeroAbundance fullVersion  
0         False       False  
1         False       False  
2         False       False  
3         False       False  
4         False       False  

[5 rows x 24 columns]

Checking the roost data and loading it into a dataframe.

In [6]:
!head -n 5 'data/RoostObservations.csv'

observationID,recordKey,organisationName,datasetKey,surveyKey,sampleKey,gridReference,precision,siteKey,siteName,featureKey,startDate,endDate,dateType,recorder,determiner,pTaxonVersionKey,taxonName,authority,commonName,taxonGroup,sensitive,zeroAbundance,fullVersion,Abundance,Comments,HRGridReference,LocationName,RecordType,SpeciesIDMethod,SurveyType
463185100,7871,The Bat Conservation Trust,GA000616,1,1,NN5620,1km,,,129467,16/06/2007,16/06/2007,Day,,,NBNSYS0000005102,Plecotus auritus,"(Linnaeus, 1758)",Brown Long-eared Bat,terrestrial mammal,false,false,false,,,,,,,
463185128,1506,The Bat Conservation Trust,GA000616,1,1,NR2767,1km,,,130452,29/05/2010,29/05/2010,Day,,,NHMSYS0020001355,Pipistrellus pipistrellus sensu stricto,"(Schreber, 1774)",Common Pipistrelle,terrestrial mammal,false,false,false,,,,,,,
463185129,1507,The Bat Conservation Trust,GA000616,1,1,NR2767,1km,,,130452,06/06/2010,06/06/2010,Day,,,NHMSYS0020001355,Pipistrellus pipistrellus sensu stricto,"(Schreber, 1774)",Common

In [7]:
!tail -n 5 'data/RoostObservations.csv'

463209487,13191,The Bat Conservation Trust,GA000616,1,1,ST1821,1km,,,653453,15/06/2011,15/06/2011,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false,,,,,,,
463209488,1738,The Bat Conservation Trust,GA000616,1,1,ST1821,1km,,,653453,29/06/2010,29/06/2010,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false,,,,,,,
463209489,1739,The Bat Conservation Trust,GA000616,1,1,ST1821,1km,,,653453,30/06/2010,30/06/2010,Day,,,NHMSYS0000528030,Myotis mystacinus/brandtii,null,Whiskered/Brandt's Bat,terrestrial mammal,false,false,false,,,,,,,
463209975,21920,The Bat Conservation Trust,GA000616,1,1,SX8166,1km,,,665485,07/06/2004,07/06/2004,Day,,,NHMSYS0000080177,Rhinolophus hipposideros,"(Bechstein, 1800)",Lesser Horseshoe Bat,terrestrial mammal,false,false,false,,,,,,,
463209976,21921,The Bat Conservation Trust,GA000616,1,1,SX8166,1km,,,665485,17/06/2004,17/06/2004,Day,,,NHMSYS0

In [8]:
roost_df = pd.read_csv('data/RoostObservations.csv', engine='python')

In [9]:
roost_df.head()

observationID  recordKey            organisationName datasetKey  surveyKey  \
0      463185100       7871  The Bat Conservation Trust   GA000616          1   
1      463185128       1506  The Bat Conservation Trust   GA000616          1   
2      463185129       1507  The Bat Conservation Trust   GA000616          1   
3      463185205        756  The Bat Conservation Trust   GA000616          1   
4      463185255      12799  The Bat Conservation Trust   GA000616          1   

   sampleKey gridReference precision  siteKey  siteName    ...      sensitive  \
0          1        NN5620       1km      NaN       NaN    ...          False   
1          1        NR2767       1km      NaN       NaN    ...          False   
2          1        NR2767       1km      NaN       NaN    ...          False   
3          1        TA0333       1km      NaN       NaN    ...          False   
4          1        NN7102       1km      NaN       NaN    ...          False   

  zeroAbundance fullVersion Abundance  Comments  HRGridReference LocationName  \
0         False       False       NaN       NaN              NaN          NaN   
1         False       False       NaN       NaN              NaN          NaN   
2         False       False       NaN       NaN              NaN          NaN   
3         False       False       NaN       NaN              NaN          NaN   
4         False       False       NaN       NaN              NaN          NaN   

  RecordType SpeciesIDMethod SurveyType  
0        NaN             NaN        NaN  
1        NaN             NaN        NaN  
2        NaN             NaN        NaN  
3        NaN             NaN        NaN  
4        NaN             NaN        NaN  

[5 rows x 31 columns]

# Exploring the datasets

In [10]:
hibernation_df.columns

Index(['observationID', 'recordKey', 'organisationName', 'datasetKey',
       'surveyKey', 'sampleKey', 'gridReference', 'precision', 'siteKey',
       'siteName', 'featureKey', 'startDate', 'endDate', 'dateType',
       'recorder', 'determiner', 'pTaxonVersionKey', 'taxonName', 'authority',
       'commonName', 'taxonGroup', 'sensitive', 'zeroAbundance',
       'fullVersion'],
      dtype='object')

In [11]:
roost_df.columns

Index(['observationID', 'recordKey', 'organisationName', 'datasetKey',
       'surveyKey', 'sampleKey', 'gridReference', 'precision', 'siteKey',
       'siteName', 'featureKey', 'startDate', 'endDate', 'dateType',
       'recorder', 'determiner', 'pTaxonVersionKey', 'taxonName', 'authority',
       'commonName', 'taxonGroup', 'sensitive', 'zeroAbundance', 'fullVersion',
       'Abundance', 'Comments', 'HRGridReference', 'LocationName',
       'RecordType', 'SpeciesIDMethod', 'SurveyType'],
      dtype='object')

In [12]:
hibernation_df.dtypes

observationID         int64
recordKey             int64
organisationName     object
datasetKey           object
surveyKey             int64
sampleKey             int64
gridReference        object
precision            object
siteKey             float64
siteName            float64
featureKey            int64
startDate            object
endDate              object
dateType             object
recorder            float64
determiner          float64
pTaxonVersionKey     object
taxonName            object
authority            object
commonName           object
taxonGroup           object
sensitive              bool
zeroAbundance          bool
fullVersion            bool
dtype: object

In [13]:
roost_df.dtypes

observationID         int64
recordKey             int64
organisationName     object
datasetKey           object
surveyKey             int64
sampleKey             int64
gridReference        object
precision            object
siteKey             float64
siteName            float64
featureKey            int64
startDate            object
endDate              object
dateType             object
recorder            float64
determiner          float64
pTaxonVersionKey     object
taxonName            object
authority            object
commonName           object
taxonGroup           object
sensitive              bool
zeroAbundance          bool
fullVersion            bool
Abundance           float64
Comments            float64
HRGridReference     float64
LocationName        float64
RecordType          float64
SpeciesIDMethod     float64
SurveyType          float64
dtype: object

In [14]:
set(roost_df.columns) - set(hibernation_df.columns)

{'Abundance',
 'Comments',
 'HRGridReference',
 'LocationName',
 'RecordType',
 'SpeciesIDMethod',
 'SurveyType'}

# The two datasets do not contain the same number of columns 

Do the additional columns in the roost dataset contain anything useful?

In [15]:
roost_df.Abundance.unique()

array([ nan])

In [16]:
roost_df.Comments.unique()

array([ nan])

In [17]:
roost_df.HRGridReference.unique()

array([ nan])

In [18]:
roost_df.LocationName.unique()

array([ nan])

In [19]:
roost_df.RecordType.unique()

array([ nan])

In [20]:
roost_df.SpeciesIDMethod.unique()

array([ nan])

In [21]:
roost_df.SurveyType.unique()

array([ nan])

So all of the additional columns in the roost dataset contain NaN values and as such are not useful.

Columns that both datasets have in common.

In [22]:
set(roost_df.columns).intersection(hibernation_df.columns)

{'authority',
 'commonName',
 'datasetKey',
 'dateType',
 'determiner',
 'endDate',
 'featureKey',
 'fullVersion',
 'gridReference',
 'observationID',
 'organisationName',
 'pTaxonVersionKey',
 'precision',
 'recordKey',
 'recorder',
 'sampleKey',
 'sensitive',
 'siteKey',
 'siteName',
 'startDate',
 'surveyKey',
 'taxonGroup',
 'taxonName',
 'zeroAbundance'}

In [23]:
roost_df.describe

<bound method NDFrame.describe of        observationID  recordKey            organisationName datasetKey  \
0          463185100       7871  The Bat Conservation Trust   GA000616   
1          463185128       1506  The Bat Conservation Trust   GA000616   
2          463185129       1507  The Bat Conservation Trust   GA000616   
3          463185205        756  The Bat Conservation Trust   GA000616   
4          463185255      12799  The Bat Conservation Trust   GA000616   
5          463185256      12800  The Bat Conservation Trust   GA000616   
6          463185451        690  The Bat Conservation Trust   GA000616   
7          463185492      17833  The Bat Conservation Trust   GA000616   
8          463185571       5151  The Bat Conservation Trust   GA000616   
9          463185572       5152  The Bat Conservation Trust   GA000616   
10         463185573       5153  The Bat Conservation Trust   GA000616   
11         463185574       5154  The Bat Conservation Trust   GA000616   
12  

In [24]:
hibernation_df.describe

<bound method NDFrame.describe of        observationID  recordKey            organisationName datasetKey  \
0          463166539       1767  The Bat Conservation Trust   GA000612   
1          463166623       4325  The Bat Conservation Trust   GA000612   
2          463166624       4326  The Bat Conservation Trust   GA000612   
3          463166625       4327  The Bat Conservation Trust   GA000612   
4          463166626       4328  The Bat Conservation Trust   GA000612   
5          463166627       4329  The Bat Conservation Trust   GA000612   
6          463166628       4330  The Bat Conservation Trust   GA000612   
7          463166629       4331  The Bat Conservation Trust   GA000612   
8          463166630       4332  The Bat Conservation Trust   GA000612   
9          463166631       4333  The Bat Conservation Trust   GA000612   
10         463166743      23494  The Bat Conservation Trust   GA000612   
11         463166262      17276  The Bat Conservation Trust   GA000612   
12  

Most of the above columns in both datasets seems to repeat the same information for each row and even those that do not seem to have a lot of NaN values.

Each column is for a single observation of a bat, as such some columns do not seem very useful.

# Exploring which common columns are useful for comparing the datasets

The authority columns:

In [25]:
hibernation_df.authority.unique()

array(['(Leach, 1825)', '(Schreber, 1774)', '(Eversmann, 1845)',
       '(Linnaeus, 1758)', '(Kuhl, 1817)', 'Blumenbach, 1779',
       '(Fischer, 1829)', 'Geoffroy, 1818', '(Bechstein, 1800)', 'null',
       '(Borkhausen, 1797)', 'Kaup, 1829'], dtype=object)

In [26]:
roost_df.authority.unique()

array(['(Linnaeus, 1758)', '(Schreber, 1774)', '(Leach, 1825)',
       '(Kuhl, 1817)', '(Bechstein, 1800)', 'Blumenbach, 1779',
       '(Eversmann, 1845)', 'null', '(Fischer, 1829)'], dtype=object)

The authority columns have values which are not shared. This column doesn't appear to be useful anyway as it appears to say who named the bat species.

The dateType colulmns:

In [27]:
hibernation_df.dateType.unique()

array(['Day'], dtype=object)

In [28]:
roost_df.dateType.unique()

array(['Day', 'Year'], dtype=object)

In [29]:
roost_df[roost_df['dateType']=='Day'].count()
roost_df[roost_df['dateType']=='Year'].count()
#dateType value of 'Day' 22360 records
#dateType value of 'Year' 1319 records
roost_df[roost_df['dateType']=='Year'].head()

observationID  recordKey            organisationName datasetKey  \
435      463194934    1009423  The Bat Conservation Trust   GA000616   
441      463195128    1001226  The Bat Conservation Trust   GA000616   
474      463196060    1015655  The Bat Conservation Trust   GA000616   
581      463196865    1006994  The Bat Conservation Trust   GA000616   
636      463197217    1000063  The Bat Conservation Trust   GA000616   

     surveyKey  sampleKey gridReference precision  siteKey  siteName  \
435          1          1        SE1314       1km      NaN       NaN   
441          1          1        NT5977       1km      NaN       NaN   
474          1          1        TA0333       1km      NaN       NaN   
581          1          1        NS8992       1km      NaN       NaN   
636          1          1        NM6863       1km      NaN       NaN   

       ...      sensitive zeroAbundance fullVersion Abundance  Comments  \
435    ...          False         False       False       NaN       NaN   
441    ...          False         False       False       NaN       NaN   
474    ...          False         False       False       NaN       NaN   
581    ...          False         False       False       NaN       NaN   
636    ...          False         False       False       NaN       NaN   

     HRGridReference LocationName RecordType SpeciesIDMethod SurveyType  
435              NaN          NaN        NaN             NaN        NaN  
441              NaN          NaN        NaN             NaN        NaN  
474              NaN          NaN        NaN             NaN        NaN  
581              NaN          NaN        NaN             NaN        NaN  
636              NaN          NaN        NaN             NaN        NaN  

[5 rows x 31 columns]

In [30]:
roost_df[roost_df['dateType']=='Day'].head()

observationID  recordKey            organisationName datasetKey  surveyKey  \
0      463185100       7871  The Bat Conservation Trust   GA000616          1   
1      463185128       1506  The Bat Conservation Trust   GA000616          1   
2      463185129       1507  The Bat Conservation Trust   GA000616          1   
3      463185205        756  The Bat Conservation Trust   GA000616          1   
4      463185255      12799  The Bat Conservation Trust   GA000616          1   

   sampleKey gridReference precision  siteKey  siteName    ...      sensitive  \
0          1        NN5620       1km      NaN       NaN    ...          False   
1          1        NR2767       1km      NaN       NaN    ...          False   
2          1        NR2767       1km      NaN       NaN    ...          False   
3          1        TA0333       1km      NaN       NaN    ...          False   
4          1        NN7102       1km      NaN       NaN    ...          False   

  zeroAbundance fullVersion Abundance  Comments  HRGridReference LocationName  \
0         False       False       NaN       NaN              NaN          NaN   
1         False       False       NaN       NaN              NaN          NaN   
2         False       False       NaN       NaN              NaN          NaN   
3         False       False       NaN       NaN              NaN          NaN   
4         False       False       NaN       NaN              NaN          NaN   

  RecordType SpeciesIDMethod SurveyType  
0        NaN             NaN        NaN  
1        NaN             NaN        NaN  
2        NaN             NaN        NaN  
3        NaN             NaN        NaN  
4        NaN             NaN        NaN  

[5 rows x 31 columns]

In [31]:
roost_df[['startDate', 'endDate']].head()

startDate     endDate
0  16/06/2007  16/06/2007
1  29/05/2010  29/05/2010
2  06/06/2010  06/06/2010
3  13/07/2010  13/07/2010
4  11/07/2010  11/07/2010

In [32]:
roost_df_year = roost_df[roost_df['dateType']=='Year']

In [33]:
roost_df_year[['startDate', 'endDate']].head()

startDate     endDate
435  01/01/2002  31/12/2002
441  01/01/2004  31/12/2004
474  01/01/2013  31/12/2013
581  01/01/2012  31/12/2012
636  01/01/2002  31/12/2002

The hibernation dataset has a dateType of day while the roost dataset has day and year. 

There doesn't appear to be anything different about the values that have dataType year and roost metadata states that roost observations are for two days in summer. So the dateType column does not appear to be useful.

Examining the other date columns.

In [34]:
len(hibernation_df.endDate.unique())

2180

In [35]:
len(hibernation_df.startDate.unique())

2180

There are 2180 unique startDate and endDate values in the hibernation dataset.

In [36]:
len(roost_df.endDate.unique())

1930

In [37]:
len(roost_df.startDate.unique())

1930

There are 1930 unique startDate and endDate values in the roost dataset.

The determiner columns:

In [38]:
hibernation_df.determiner.unique()

array([ nan])

In [39]:
roost_df.determiner.unique()

array([ nan])

The determiner column is NaN for both datasets so does not appear to be useful.

The featureKey columns:

In [40]:
#hibernation_df.featureKey.unique()

In [41]:
roost_df.featureKey.unique()

array([129467, 130452, 160033, ..., 849696, 811992, 665485], dtype=int64)

It is not clear what the featureKey column is for but it does not seem useful.

The fullVersion columns:

In [42]:
hibernation_df.fullVersion.unique()

array([False], dtype=object)

In [43]:
roost_df.fullVersion.unique()

array([False], dtype=object)

Some of the data is restricted due to vulnerable species and to protect the privacy of land owners as stated in the metadata. So it seems this column just states that it is not the full version of the data.

The gridRefernce columns:

In [44]:
len(hibernation_df.gridReference.unique())

350

In [45]:
len(roost_df.gridReference.unique())

2026

Each gridReference is for a unique site and there are 350 in the hibernation dataset and 2026 in the rooost dataset.

The observationID columns:

In [46]:
len(hibernation_df.observationID.unique())

22899

In [47]:
len(roost_df.observationID.unique())

23679

Each observation of a bat has an observationID so there are 22,899 bats recorded in the hibernation data and 23,679 in the roost data.

The organisationName columns:

In [48]:
hibernation_df.organisationName.unique()

array(['The Bat Conservation Trust'], dtype=object)

In [49]:
roost_df.organisationName.unique()

array(['The Bat Conservation Trust'], dtype=object)

Both datasets have unsurprisingly  the same value repeated for the organisationName column so it is not useful.

The pTaxonVersionKey columns:

In [50]:
hibernation_df.pTaxonVersionKey.unique()

array(['NBNSYS0100004720', 'NHMSYS0020001355', 'NHMSYS0000528026',
       'NHMSYS0000080178', 'NHMSYS0020001356', 'NBNSYS0000005102',
       'NHMSYS0000528028', 'NHMSYS0000376160', 'NHMSYS0000528008',
       'NHMSYS0000080176', 'NBNSYS0000005103', 'NBNSYS0000136687',
       'NHMSYS0000080177', 'NHMSYS0000528024', 'NHMSYS0000528030',
       'NHMSYS0000080184', 'NHMSYS0000080182', 'NHMSYS0000080185',
       'NHMSYS0000080183', 'NBNSYS0000040730', 'NHMSYS0000080186'], dtype=object)

In [51]:
hibernation_df.pTaxonVersionKey.unique()

array(['NBNSYS0100004720', 'NHMSYS0020001355', 'NHMSYS0000528026',
       'NHMSYS0000080178', 'NHMSYS0020001356', 'NBNSYS0000005102',
       'NHMSYS0000528028', 'NHMSYS0000376160', 'NHMSYS0000528008',
       'NHMSYS0000080176', 'NBNSYS0000005103', 'NBNSYS0000136687',
       'NHMSYS0000080177', 'NHMSYS0000528024', 'NHMSYS0000528030',
       'NHMSYS0000080184', 'NHMSYS0000080182', 'NHMSYS0000080185',
       'NHMSYS0000080183', 'NBNSYS0000040730', 'NHMSYS0000080186'], dtype=object)

Seems to be the taxon version that is being used. I don't think this will be useful.

# Differences in precision values

The precision columns:

In [52]:
hibernation_df.precision.unique()

array(['10km'], dtype=object)

In [53]:
roost_df.precision.unique()

array(['1km', '10km'], dtype=object)

In [54]:
roost_df.gridReference[roost_df['precision']=='1km'].count() 

23674

In [55]:
roost_df.gridReference[roost_df['precision']=='10km'].count()

5

The roost data has 23,674 observations at 1km and 5 observations at 10km.

In [56]:
roost_df.gridReference[roost_df['precision']=='10km']

121     ST76
122     ST76
1074    SE30
1075    SE30
1076    SE30
Name: gridReference, dtype: object

Only two sites are at a 10km precision.

The precision to which the data is given is restricted due to sensitive species and privacy reasons as discussed above.

Given the above stated low number of values I have elected to keep the data that is at a 10km precision.

The recordKey columns:

In [57]:
len(hibernation_df.recordKey.unique())

22899

In [58]:
len(roost_df.recordKey.unique())

23679

Both datasets have a unique record key for each row.

The recorder columns:

In [59]:
hibernation_df.recorder.unique()

array([ nan])

In [60]:
roost_df.recorder.unique()

array([ nan])

This is likely the name of the person who recorded the observation either way it is a NaN value in both datasets and therefore not useful.

The sampleKey columns:

In [61]:
hibernation_df.sampleKey.unique()

array([1], dtype=int64)

In [62]:
roost_df.sampleKey.unique()

array([1], dtype=int64)

The sampleKey columns just repeat the same value and as such are not useful.

The sensitive columns:

In [63]:
hibernation_df.sensitive.unique()

array([False], dtype=object)

In [64]:
roost_df.sensitive.unique()

array([False], dtype=object)

Just indicates that this dataset does not contain sensitive information as this is the restricted dataset. 

The siteName columns:

In [65]:
hibernation_df.siteName.unique()

array([    nan,  10032.,  10461.,  10393.,  10431.,  10394.,  30144.,
        10030.,  10131.,  20036.,  10374.,  10357.,  10436.,  10056.,
        10238.,  20003.,  52056.,  30268.,  10132.,  10389.,  10353.,
        10355.,  20001.,  10421.,  10058.,  10536.,  10483.,  10484.,
        10535.,  30338.,  10359.,  30062.,  30118.,  30126.,  10057.,
        10059.,  30350.,  30351.,  10397.,  10481.,  10513.,  10534.,
        30420.,  30319.,  30096.,  30170.,  30072.,  10053.,  10139.,
        30213.,  30340.,  10054.,  30067.,  10212.,  10453.,  20002.,
        30285.,  10140.,  30328.,  10060.,  10055.,  10136.,  30054.,
        30065.,  30039.,  30066.,  10153.,  30008.,  30344.,  10352.,
        20042.,  10324.,  10354.,  10560.,  30394.,  10563.])

In [66]:
len(hibernation_df.siteName.unique())

76

In [67]:
hibernation_df.siteName.unique()

array([    nan,  10032.,  10461.,  10393.,  10431.,  10394.,  30144.,
        10030.,  10131.,  20036.,  10374.,  10357.,  10436.,  10056.,
        10238.,  20003.,  52056.,  30268.,  10132.,  10389.,  10353.,
        10355.,  20001.,  10421.,  10058.,  10536.,  10483.,  10484.,
        10535.,  30338.,  10359.,  30062.,  30118.,  30126.,  10057.,
        10059.,  30350.,  30351.,  10397.,  10481.,  10513.,  10534.,
        30420.,  30319.,  30096.,  30170.,  30072.,  10053.,  10139.,
        30213.,  30340.,  10054.,  30067.,  10212.,  10453.,  20002.,
        30285.,  10140.,  30328.,  10060.,  10055.,  10136.,  30054.,
        30065.,  30039.,  30066.,  10153.,  30008.,  30344.,  10352.,
        20042.,  10324.,  10354.,  10560.,  30394.,  10563.])

In [68]:
len(roost_df.siteName.unique())

200

It seems likely that the siteName column contains a key for each site and that some of them have been removed. But either way this column doesn't seem useful. 

Furthermore I have decided to use grid reference as the basis for differentiating sites as this is a value that all observations in both datasets have.

The surveyKey columns:

In [69]:
hibernation_df.surveyKey.unique()

array([1], dtype=int64)

In [70]:
roost_df.surveyKey.unique()

array([1], dtype=int64)

Same value repeated, not useful.

The taxonGroup columns:

In [71]:
hibernation_df.taxonGroup.unique()

array(['terrestrial mammal'], dtype=object)

In [72]:
roost_df.taxonGroup.unique()

array(['terrestrial mammal'], dtype=object)

Same value repeated, not useful.

# Name columns

The commonName columns:

In [73]:
hibernation_df.commonName.unique()

array(['Soprano Pipistrelle', 'Common Pipistrelle', "Brandt's Bat",
       'Western Barbastelle', 'Pipistrelle', 'Brown Long-eared Bat',
       "Daubenton's Bat", 'Bat', 'Serotine', 'Greater Horseshoe Bat',
       'Grey Long-eared Bat', 'Long-eared Bat species',
       'Lesser Horseshoe Bat', "Bechstein's Bat", "Whiskered/Brandt's Bat",
       "Natterer's Bat", 'Mouse-eared Bat', 'Lesser Noctule',
       'Whiskered Bat', 'Myotis Bat species', 'Noctule Bat'], dtype=object)

In [74]:
roost_df.commonName.unique()

array(['Brown Long-eared Bat', 'Common Pipistrelle', 'Pipistrelle',
       'Soprano Pipistrelle', "Natterer's Bat", 'Lesser Horseshoe Bat',
       'Serotine', 'Greater Horseshoe Bat', 'Noctule Bat', 'Whiskered Bat',
       'Bat', "Daubenton's Bat", 'Lesser Noctule', "Brandt's Bat",
       "Whiskered/Brandt's Bat", 'Grey Long-eared Bat', "Bechstein's Bat"], dtype=object)

# Differences in name columns

The commonName columns contains for each respective dataset do not contain the same number of values. I think that taxonName may be a more appropriate choice for identifying the bat species.

In [75]:
hibernation_df.taxonName.unique()

array(['Pipistrellus pygmaeus', 'Pipistrellus pipistrellus sensu stricto',
       'Myotis brandtii', 'Barbastella barbastellus',
       'Pipistrellus pipistrellus sensu lato', 'Plecotus auritus',
       'Myotis daubentonii', 'Chiroptera', 'Eptesicus serotinus',
       'Rhinolophus ferrumequinum', 'Plecotus austriacus', 'Plecotus',
       'Rhinolophus hipposideros', 'Myotis bechsteinii',
       'Myotis mystacinus/brandtii', 'Myotis nattereri', 'Myotis myotis',
       'Nyctalus leisleri', 'Myotis mystacinus', 'Myotis',
       'Nyctalus noctula'], dtype=object)

In [76]:
roost_df.taxonName.unique()

array(['Plecotus auritus', 'Pipistrellus pipistrellus sensu stricto',
       'Pipistrellus pipistrellus sensu lato', 'Pipistrellus pygmaeus',
       'Myotis nattereri', 'Rhinolophus hipposideros',
       'Eptesicus serotinus', 'Rhinolophus ferrumequinum',
       'Nyctalus noctula', 'Myotis mystacinus', 'Chiroptera',
       'Myotis daubentonii', 'Nyctalus leisleri', 'Myotis brandtii',
       'Myotis mystacinus/brandtii', 'Plecotus austriacus',
       'Myotis bechsteinii'], dtype=object)

So it seems that the taxonName columns also contain a different number of values.

Why these bat species are present in the hibernation data but not the roost data is unclear. 

For the purpose of comparing the two datasets having bat species present in one and not the other is not useful. So the data for bat species that are not present in both data sets will be removed after shaping the tables.

The zeroAbundance columns:

In [77]:
hibernation_df.zeroAbundance.unique()

array([False], dtype=object)

In [78]:
roost_df.zeroAbundance.unique()

array([False], dtype=object)

Each row in the datasets is for a specific bat observation so natrually zero abundance is always false. So this column is not very useful.

The dataset key columns.

In [79]:
hibernation_df.datasetKey.unique()

array(['GA000612'], dtype=object)

In [80]:
roost_df.datasetKey.unique()

array(['GA000616'], dtype=object)

# Changing the data type of the date values for each dataset

In [81]:
hibernation_df['startDate'] = pd.to_datetime(hibernation_df['startDate'])
hibernation_df['endDate'] = pd.to_datetime(hibernation_df['endDate'])
hibernation_df['year'] = hibernation_df['endDate'].dt.year

In [82]:
roost_df['startDate'] = pd.to_datetime(roost_df['startDate'])
roost_df['endDate'] = pd.to_datetime(roost_df['endDate'])
roost_df['year'] = roost_df['endDate'].dt.year

# Shaping the dataframes by selecting the useful columns

In [83]:
columns = hibernation_df.columns.tolist()
columns = ['observationID', 'gridReference', 'year', 'taxonName']
columns

hibernation_short_df = hibernation_df[columns]

hibernation_short_df.tail(5)

observationID gridReference  year                   taxonName
22894      463184995          TQ35  2013  Myotis mystacinus/brandtii
22895      463185027          TQ21  2005  Myotis mystacinus/brandtii
22896      463185029          TQ21  2005  Myotis mystacinus/brandtii
22897      463185031          TQ31  2005  Myotis mystacinus/brandtii
22898      463185032          TQ31  2008  Myotis mystacinus/brandtii

In [84]:
roost_df.head()

columns = roost_df.columns.tolist()
columns = ['observationID', 'gridReference', 'year', 'taxonName']
columns

roost_short_df = roost_df[columns]
roost_short_df.tail(5)

observationID gridReference  year                   taxonName
23674      463209487        ST1821  2011  Myotis mystacinus/brandtii
23675      463209488        ST1821  2010  Myotis mystacinus/brandtii
23676      463209489        ST1821  2010  Myotis mystacinus/brandtii
23677      463209975        SX8166  2004    Rhinolophus hipposideros
23678      463209976        SX8166  2004    Rhinolophus hipposideros

# Removing taxonName values from the hibernation dataset that are not also in the roost dataset

In [85]:
to_delete = set(hibernation_short_df['taxonName'].unique().tolist()) - set(roost_short_df['taxonName'].unique().tolist())
to_delete

{'Barbastella barbastellus', 'Myotis', 'Myotis myotis', 'Plecotus'}

In [86]:
hibernation_reduced_df = hibernation_short_df[~hibernation_short_df['taxonName'].isin(to_delete)]

Checking that the number of bat species in the hibernation dataset is now 17.

In [87]:
len(hibernation_reduced_df['taxonName'].unique())

17

# Time factors

In [88]:
hibernation_reduced_df['year'].sort_values().unique()

array([1964, 1965, 1966, 1967, 1968, 1969, 1970, 1972, 1973, 1974, 1975,
       1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986,
       1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997,
       1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008,
       2009, 2010, 2011, 2012, 2013, 2014], dtype=int64)

In [89]:
roost_short_df['year'].sort_values().unique()

array([1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013, 2014], dtype=int64)

The hibernation datasets covers a longer period of time. I'm not sure the difference in time span is significant and in any case the data is for different areas.

So far I have elected to retain the data for the hibernation sites that predates that for roost sites. I have done this because both data sets start out with very few observations.

The risk of retaining the data for hibernation sites is that there could be a temporal skew in the data. That is there may have been historical factors that affected certain species of bats disproportionately at different times. 

But this is speculation on my part and I still believe that retaining the earlier data in the hibernation set will provide a better average picture of species abundance relative to roost sites.

Furthermore both data sets also have a similar number of bats 22,139 for hibernation sites vs 23,679 for roost sites despite the larger area in the roost dataset.

# How do the two datasets compare over time?

In [90]:
roost_by_year = roost_short_df.groupby(['year'])['taxonName'].count()
roost_by_year_df = roost_by_year.to_frame()
type(roost_by_year_df)
roost_by_year_df = roost_by_year_df.rename(columns = {'taxonName': 'roost_abundance'})
#roost_by_year_df

In [91]:
hibernation_by_year = hibernation_short_df.groupby(['year'])['taxonName'].count()
hibernation_by_year_df = hibernation_by_year.to_frame()
hibernation_by_year_df = hibernation_by_year_df.rename(columns = {'taxonName': 'hibernation_abundance'})
#hibernation_by_year_df

In [92]:
combined_abundance_year = pd.concat([hibernation_by_year_df, roost_by_year_df], axis=1, join='outer')
combined_abundance_year.transpose()

year                   1964  1965  1966  1967  1968  1969  1970  1972  1973  \
hibernation_abundance   4.0   5.0   4.0   6.0   4.0   4.0   5.0   4.0   5.0   
roost_abundance         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

year                   1974   ...      2005    2006    2007    2008    2009  \
hibernation_abundance   9.0   ...    1011.0  1206.0  1133.0  1313.0  1336.0   
roost_abundance         NaN   ...    1306.0  1377.0  1392.0  1498.0  1555.0   

year                     2010    2011    2012    2013    2014  
hibernation_abundance  1452.0  1427.0  1402.0  1451.0  1119.0  
roost_abundance        1519.0  1461.0  1274.0  1368.0  1362.0  

[2 rows x 50 columns]

In [93]:
combined_abundance_year.plot(figsize=(18,10))

The two lines show a similar rise in the number of bats over time although with variations in their relative shapes.

A noticeable factor is that the roost data despite being for a shorter time span than the hibernation data actually has more recorded bats. This is likely due to the roost data covering more sites and a larger geographical area.

How about the variation in species abundance in the two datasets? Is the composition of hibernation sites different than roost sites?

In [94]:
hs_df = hibernation_reduced_df.groupby(['taxonName'])['taxonName'].count()

In [95]:
hibernation_count = hs_df.to_frame()

In [96]:
rs_df = roost_short_df.groupby(['taxonName'])['taxonName'].count()

In [97]:
roost_count = rs_df.to_frame()

In [98]:
roost_count = roost_count.rename(columns = {'taxonName': 'roost_abundance'})
#roost_count

In [99]:
hibernation_count = hibernation_count.rename(columns = {'taxonName': 'hibernation_abundance'})
#hibernation_count

In [100]:
combined_abundance = pd.concat([hibernation_count, roost_count], axis=1, join='outer')
combined_abundance

hibernation_abundance  \
taxonName                                                        
Chiroptera                                                 557   
Eptesicus serotinus                                         14   
Myotis bechsteinii                                          75   
Myotis brandtii                                            135   
Myotis daubentonii                                        3945   
Myotis mystacinus                                          204   
Myotis mystacinus/brandtii                                1323   
Myotis nattereri                                          5056   
Nyctalus leisleri                                            6   
Nyctalus noctula                                             7   
Pipistrellus pipistrellus sensu lato                       412   
Pipistrellus pipistrellus sensu stricto                    109   
Pipistrellus pygmaeus                                       21   
Plecotus auritus                                          2805   
Plecotus austriacus                                          1   
Rhinolophus ferrumequinum                                 2963   
Rhinolophus hipposideros                                  4506   

                                         roost_abundance  
taxonName                                                 
Chiroptera                                            54  
Eptesicus serotinus                                 1107  
Myotis bechsteinii                                    22  
Myotis brandtii                                       60  
Myotis daubentonii                                   160  
Myotis mystacinus                                     97  
Myotis mystacinus/brandtii                            33  
Myotis nattereri                                    1040  
Nyctalus leisleri                                     11  
Nyctalus noctula                                      72  
Pipistrellus pipistrellus sensu lato                3478  
Pipistrellus pipistrellus sensu stricto             6475  
Pipistrellus pygmaeus                               4421  
Plecotus auritus                                    1934  
Plecotus austriacus                                    5  
Rhinolophus ferrumequinum                            514  
Rhinolophus hipposideros                            4196

In [101]:
combined_abundance.plot(kind='bar', figsize=(18,10))
plt.xlabel('Taxon name')
plt.ylabel('Number of bats')

The graph above shows a marked difference in species abundance between the hibernation and roost datasets.

So it appears that the composition of roost and hibernation sites is different.

# But is the difference in species abundance between roost and hibernations sites significant?

The null hypothesis is that it is not any statistically significant difference between roost and hibernation site bat species composition.

This will be tested using Chi-squared to see if the difference between hibernation an roost site abundance for the various bat species are significant.

If the result is not significant then the P-value should be one, meaning that there is no difference between expected and actual abundance.

# Calculating expected values and P value

Generating values that would be expected if the abundances were evenly distributed

In [102]:
#code adapted from notebook 14.3
expected_results = pd.DataFrame({c: {r: (combined_abundance[c].sum() * 
                                                combined_abundance.loc[r].sum() / 
                                                combined_abundance.sum().sum() )
                  for r in combined_abundance[c].index} 
              for c in combined_abundance}
)
expected_results

hibernation_abundance  \
Chiroptera                                          295.231765   
Eptesicus serotinus                                 541.660897   
Myotis bechsteinii                                   46.869855   
Myotis brandtii                                      94.222904   
Myotis daubentonii                                 1983.512921   
Myotis mystacinus                                   145.441508   
Myotis mystacinus/brandtii                          655.211576   
Myotis nattereri                                   2945.552927   
Nyctalus leisleri                                     8.214304   
Nyctalus noctula                                     38.172356   
Pipistrellus pipistrellus sensu lato               1879.626129   
Pipistrellus pipistrellus sensu stricto            3181.351783   
Pipistrellus pygmaeus                              2146.349426   
Plecotus auritus                                   2289.858156   
Plecotus austriacus                                   2.899166   
Rhinolophus ferrumequinum                          1680.066851   
Rhinolophus hipposideros                           4204.757475   

                                         roost_abundance  
Chiroptera                                    315.768235  
Eptesicus serotinus                           579.339103  
Myotis bechsteinii                             50.130145  
Myotis brandtii                               100.777096  
Myotis daubentonii                           2121.487079  
Myotis mystacinus                             155.558492  
Myotis mystacinus/brandtii                    700.788424  
Myotis nattereri                             3150.447073  
Nyctalus leisleri                               8.785696  
Nyctalus noctula                               40.827644  
Pipistrellus pipistrellus sensu lato         2010.373871  
Pipistrellus pipistrellus sensu stricto      3402.648217  
Pipistrellus pygmaeus                        2295.650574  
Plecotus auritus                             2449.141844  
Plecotus austriacus                             3.100834  
Rhinolophus ferrumequinum                    1796.933149  
Rhinolophus hipposideros                     4497.242525

Applying the above actual and expected values to Chi-squared:

In [103]:
scipy.stats.chisquare(combined_abundance, expected_results, axis=None)

Power_divergenceResult(statistic=23797.463852411889, pvalue=0.0)

This produces a P value of 0.0 meaning that the difference between the composition of the roost and hibernation dataset is statistically significant.

So it seems that hibernation and roost sites do have a statistically significant different abundance of species relative to one another and so a different composition.

There is a problem with applying Chi-squared with combined values that are less than 5 as this can give unreliable results. This should not have affected the result however as only one of the categorical values was less than 5.

The result is also of limited value as it says nothing about why they are different.